plugins/upower.c

```diff
@@ -0,0 +1,478 @@
+			ofono_error("%s: 'Percentage' val != VARIANT", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_recurse(&entry, &val);
+
+		if (dbus_message_iter_get_arg_type(&val) != DBUS_TYPE_DOUBLE) {
+			ofono_error("%s: 'Percentage' val != DOUBLE", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_get_basic(&val, &percentage_val);
+		percentage_found = TRUE;
+		break;
+	}
+
+	/* Other properties can change, so don't trigger cb removal */
+	if (percentage_found == FALSE)
+		return TRUE;
+
+	return update_battery_level(percentage_val);
+}
+
+static void get_property_reply(DBusPendingCall *call, void *user_data)
+{
+	double percentage_val;
+	DBusMessageIter iter, val;
+	DBusMessage *reply;
+
+	DBG("");
+
+	reply = dbus_pending_call_steal_reply(call);
+	if (reply == NULL) {
+		ofono_error("%s: dbus_message_new_method failed", __func__);
+		goto done;
+	}
+
+	if (dbus_message_get_type(reply) == DBUS_MESSAGE_TYPE_ERROR) {
+		ofono_error("%s: ERROR reply to Get('Percentage')", __func__);
+		goto done;
+	}
+
+	if (dbus_message_iter_init(reply, &iter) == FALSE) {
+		ofono_error("%s: error initializing array iter", __func__);
+		goto done;
+	}
+
+	if (dbus_message_iter_get_arg_type(&iter) != DBUS_TYPE_VARIANT) {
+		ofono_error("%s: type != VARIANT!", __func__);
+		goto done;
+	}
+
+	dbus_message_iter_recurse(&iter, &val);
+
+	if (dbus_message_iter_get_arg_type(&val) != DBUS_TYPE_DOUBLE) {
+		ofono_error("%s: type != DOUBLE!", __func__);
+		goto done;
+	}
+
+	dbus_message_iter_get_basic(&val, &percentage_val);
+
+	if (update_battery_level(percentage_val) == FALSE)
+		goto done;
+
+	upower_battery_watch = g_dbus_add_signal_watch(connection,
+						UPOWER_SERVICE,
+						battery_device_path,
+						DBUS_INTERFACE_PROPERTIES,
+						"PropertiesChanged",
+						battery_props_changed,
+						NULL, NULL);
+done:
+	if (reply)
+		dbus_message_unref(reply);
+
+	dbus_pending_call_unref(call);
+}
+
+static void emulator_hfp_watch(struct ofono_atom *atom,
+				enum ofono_atom_watch_condition cond,
+				void *data)
+{
+	struct ofono_modem *modem = data;
+
+	if (cond == OFONO_ATOM_WATCH_CONDITION_REGISTERED) {
+		DBusMessageIter iter;
+		DBusPendingCall *call;
+		DBusMessage *msg;
+		const char *iface = UPOWER_DEVICE_INTERFACE;
+		const char *property = "Percentage";
+
+		DBG("%s: REGISTERED", __func__);
+
+		/* TODO: handle removable batteries */
+
+		modems = g_list_append(modems, modem);
+
+		if (modems->next != NULL)
+			return;
+
+		/* Query current battery value */
+		msg = dbus_message_new_method_call(UPOWER_SERVICE,
+						battery_device_path,
+						DBUS_PROPERTIES_INTERFACE,
+						"Get");
+		if (msg == NULL) {
+			ofono_error("%s: dbus_message_new_method failed",
+					__func__);
+			return;
+		}
+
+		dbus_message_iter_init_append(msg, &iter);
+		dbus_message_iter_append_basic(&iter,
+						DBUS_TYPE_STRING, &iface);
+		dbus_message_iter_append_basic(&iter,
+						DBUS_TYPE_STRING, &property);
+
+		if (!dbus_connection_send_with_reply(connection, msg,
```

Comment:
I think that the call to g_dbus_add_signal_watch() should be before getting the current value, so you get events between getting the reply to this request and the registration for changes. This would also make the function more symmetrical, as you remove the signal watch here too.